### PR TITLE
[WIP] Default Packager

### PR DESCRIPTION
**Pull request by twokul**

### CODE REVIEW

File: lib/experiments/index.js [hunk 1]
```diff
@@ -3,6 +3,7 @@ const symbol = require('../utilities/symbol');
 
 let experiments = {
   MODULE_UNIFICATION: symbol('module-unification'),
+  PACKAGING: symbol('packaging'),
```
Comment (twokul) [inline lib/experiments/index.js]: not getting used anywhere yet, but will be

File: tests/unit/broccoli/default-packager/helpers.js [hunk 2]
```diff
@@ -0,0 +1,161 @@
+'use strict';
+
+const DEFAULT_APPLICATION_SOURCE = {
+  app: {
```
Comment (twokul) [inline tests/unit/broccoli/default-packager/helpers.js]: this has gotten extremely verbose, maybe we can find a better way to work with fixtures.

File: lib/broccoli/default-packager.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+ *   assembleApplicationTests(): BroccoliTree;
+ *   package(): BroccoliTree;
+ * }
+*/
+
+function prependVendorPrefixFileToVendorJs(map, path) {
+  map[path].unshift(VENDOR_PREFIX_FILE_PATH);
+}
+
+module.exports = class DefaultPackager {
```
Comment (kratiahuja) [inline lib/broccoli/default-packager.js]: This module looks very much like ember-app file.

File: tests/unit/broccoli/default-packager/helpers.js (hunk 2 above)
Comment (twokul) [reply tests/unit/broccoli/default-packager/helpers.js]: true that! we all agree that `ember-app` is trying to do too much, the key here would be to draw a clear boundary between them

File: lib/broccoli/default-packager.js [hunk 4]
```diff
@@ -0,0 +1,408 @@
...
+
+const APP_BOOT_FILE_PATH = PACKAGER_CONSTANTS.APP_BOOT_FILE_PATH;
+const APP_PREFIX_FILE_PATH = PACKAGER_CONSTANTS.APP_PREFIX_FILE_PATH;
+const APP_SUFFIX_FILE_PATH = PACKAGER_CONSTANTS.APP_SUFFIX_FILE_PATH;
+const APP_CONFIG_FILE_PATH = PACKAGER_CONSTANTS.APP_CONFIG_FILE_PATH;
+const VENDOR_PREFIX_FILE_PATH = PACKAGER_CONSTANTS.VENDOR_PREFIX_FILE_PATH;
+const VENDOR_SUFFIX_FILE_PATH = PACKAGER_CONSTANTS.VENDOR_SUFFIX_FILE_PATH;
+const ADDONS_FOLDER_NAME = PACKAGER_CONSTANTS.ADDONS_FOLDER_NAME;
+
+const DEFAULT_CONFIG = {
```
Comment (twokul) [inline lib/broccoli/default-packager.js]: we should just move it to `PACKAGER_CONSTANTS` I think

File: lib/broccoli/default-packager.js [hunk 5]
```diff
@@ -0,0 +1,408 @@
...
+ * @param {Object} options
+ * @return {BroccoliTree} Final Broccoli tree with vendor files.
+ */
+function packageVendorJs(tree, options) {
+  let importPaths = Object.keys(options.outputFilesMap);
+
+  let packagedVendorJsTrees = importPaths.map(importPath => {
+    let vendorObject = getVendorFiles(
+      options.outputFilesMap[importPath],
+      importPath === DEFAULT_CONFIG.outputPaths.vendor.js
```
Comment (ro0gr) [inline lib/broccoli/default-packager.js]: Please take into account this fix https://github.com/ember-cli/ember-cli/pull/7519

**Comment by villander** [comment]
@twokul Do you need help with this code splitting feature? I'm here to help, anything let me know

**Comment by twokul** [comment]
@villander I'll see how we can split the work

**Comment by villander** [comment]
Thanks @twokul

File: tests/helpers/default-packager.js [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+  styles: {
+    'app.css': 'html,body{height:100%}',
+  },
+  templates: {
+    'application.hbs': '{{outlet}}',
+  },
+  config: {
+    'environment.js': `module.exports = function() { return { modulePrefix: 'the-best-app-ever' }; };`,
+  },
+  'package.json': JSON.stringify({
```
Comment (kellyselden) [inline tests/helpers/default-packager.js]: Can we trim this object to the essentials just to make it work?

**Comment by twokul** [comment]
close in favour of https://github.com/ember-cli/ember-cli/pull/7826

**Pull request closed without merging**